src/Common/TraceCollector.cpp

```diff
@@ -61,17 +62,19 @@ TraceCollector::~TraceCollector()
 
 void TraceCollector::collect(TraceType trace_type, const StackTrace & stack_trace, Int64 size)
 {
-    constexpr size_t buf_size = sizeof(char) + // TraceCollector stop flag
-        8 * sizeof(char) +                     // maximum VarUInt length for string size
-        QUERY_ID_MAX_LEN * sizeof(char) +      // maximum query_id length
-        sizeof(UInt8) +                        // number of stack frames
-        sizeof(StackTrace::FramePointers) +    // collected stack trace, maximum capacity
-        sizeof(TraceType) +                    // trace type
-        sizeof(UInt64) +                       // thread_id
-        sizeof(Int64);                         // size
+    constexpr size_t buf_size = sizeof(char) /// TraceCollector stop flag
+        + sizeof(UInt8)                      /// String size
+        + QUERY_ID_MAX_LEN                   /// Maximum query_id length
+        + sizeof(UInt8)                      /// Number of stack frames
+        + sizeof(StackTrace::FramePointers)  /// Collected stack trace, maximum capacity
+        + sizeof(TraceType)                  /// trace type
+        + sizeof(UInt64)                     /// thread_id
+        + sizeof(Int64);                     /// size
+
     /// Write should be atomic to avoid overlaps
     /// (since recursive collect() is possible)
-    static_assert(buf_size < PIPE_BUF, "Only write of PIPE_BUF to pipe is atomic");
+    static_assert(PIPE_BUF >= 512);
+    static_assert(buf_size <= 512, "Only write of PIPE_BUF to pipe is atomic and the minimal known PIPE_BUF across supported platforms is 512");
```

Comment:
It is better to always compare with the minimal PIPE_BUF, so the code will not compile on Linux in the same way as on Mac.